File: examples/sample-code.java
// Sample Java code for testing the code indexer

package com.example;

import java.util.*;

// User class
public class User {
    private int id;
    private String name;
    private String email;
    private boolean isActive;

    // Constructor
    public User(int id, String name, String email) {
        this.id = id;
        this.name = name;
        this.email = email;
        this.isActive = true;
    }

    // Getter methods
    public int getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public String getEmail() {
        return email;
    }

    public boolean isActive() {
        return isActive;
    }

    // Setter methods
    public void setId(int id) {
        this.id = id;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public void setActive(boolean active) {
        isActive = active;
    }

    // Method that calls another method
    public void activate() {
        this.isActive = true;
    }

    public void deactivate() {
        this.isActive = false;
    }

    // Method that calls another function
    public boolean isValid() {
        return UserService.validateEmail(this.email) && !this.name.isEmpty();
    }

    // Method that calls another method
    public String formatName() {
        return UserService.formatUserName(this.name);
    }

    @Override
    public String toString() {
        return "User{id=" + id + ", name='" + name + "', email='" + email + "'}";
    }
}

// UserService class
public class UserService {
    private Map<Integer, User> users;

    // Constructor
    public UserService() {
        this.users = new HashMap<>();
    }

    // Method that calls another method
    public void addUser(User user) throws IllegalArgumentException {
        if (user == null) {
            throw new IllegalArgumentException("User cannot be null");
        }
        if (users.containsKey(user.getId())) {
            throw new IllegalArgumentException("User already exists: " + user.getId());
        }
        users.put(user.getId(), user);
    }

    // Method
    public User getUser(int id) {
        return users.get(id);
    }

    // Method
    public void updateUser(User user) throws IllegalArgumentException {
        if (!users.containsKey(user.getId())) {
            throw new IllegalArgumentException("User not found: " + user.getId());
        }
        users.put(user.getId(), user);
    }

    // Method
    public void deleteUser(int id) throws IllegalArgumentException {
        User removed = users.remove(id);
        if (removed == null) {
            throw new IllegalArgumentException("User not found: " + id);
        }
    }

    // Method that calls another method
    public List<User> listActiveUsers() {
        List<User> activeUsers = new ArrayList<>();
        for (User user : users.values()) {
            if (user.isActive()) {
                activeUsers.add(user);
            }
        }
        return activeUsers;
    }

    // Method that calls another function
    public User getUserByEmail(String email) {
        for (User user : users.values()) {
            if (user.getEmail().equals(email)) {
                return user;
            }
        }
        return null;
    }

    // Static method
    public static boolean validateEmail(String email) {
        return email != null && email.contains("@") && email.contains(".");
    }

    // Static method
    public static String formatUserName(String name) {
        return name != null ? name.trim().toUpperCase() : "";
    }

    // Method that calls other methods
    public static User createUser(int id, String name, String email) throws IllegalArgumentException {
        if (!validateEmail(email)) {
            throw new IllegalArgumentException("Invalid email: " + email);
        }

        String formattedName = formatUserName(name);

        return new User(id, formattedName, email);
    }

    // Method that calls other methods
    public static void processUsers(UserService service, List<User> users) throws IllegalArgumentException {
        for (User user : users) {
            service.addUser(user);
        }
    }
}

// Interface example
public interface Validator {
    boolean validate();
}

// Implementation of interface
public class UserValidator implements Validator {
    private User user;

    public UserValidator(User user) {
        this.user = user;
    }

    @Override
    public boolean validate() {
        if (user.getName() == null || user.getName().isEmpty()) {
            return false;
        }
        return UserService.validateEmail(user.getEmail());
    }
}

// Enum example
public enum UserRole {
    ADMIN,
    USER,
    GUEST
}

// Enum with constructor
public enum UserStatus {
    ACTIVE("Active"),
    INACTIVE("Inactive"),
    SUSPENDED("Suspended");

    private final String description;

    UserStatus(String description) {
        this.description = description;
    }

    public String getDescription() {
        return description;
    }
}

// Abstract class
public abstract class BaseService {
    protected String name;

    public BaseService(String name) {
        this.name = name;
    }

    public abstract void initialize();

    public String getName() {
        return name;
    }
}

// Class extending abstract class
public class AdminService extends BaseService {
    public AdminService() {
        super("AdminService");
    }

    @Override
    public void initialize() {
        System.out.println("Initializing " + getName());
    }

    // Method specific to AdminService
    public void grantPermission(String permission) {
        System.out.println("Granting permission: " + permission);
    }
}

// Constants class
public class Constants {
    public static final int MAX_USERS = 1000;
    public static final int DEFAULT_PAGE_SIZE = 20;
    public static final String DEFAULT_ROLE = "USER";
}

// Point class for testing
public class Point {
    private double x;
    private double y;

    public Point(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    public double distance() {
        return Math.sqrt(x * x + y * y);
    }
}

// Main class
public class Main {
    public static void main(String[] args) {
        UserService service = new UserService();

        try {
            User user1 = UserService.createUser(1, "John Doe", "john@example.com");
            User user2 = UserService.createUser(2, "Jane Smith", "jane@example.com");

            service.addUser(user1);
            service.addUser(user2);

            List<User> activeUsers = service.listActiveUsers();
            System.out.println("Active users: " + activeUsers.size());
        } catch (IllegalArgumentException e) {
            System.err.println("Error: " + e.getMessage());
        }
    }
}

File: examples/sample-code.rs
// Sample Rust code for testing the code indexer

// Module declaration
mod example;

// Use statements
use std::collections::HashMap;

// Constants
const MAX_USERS: usize = 1000;
const DEFAULT_PAGE_SIZE: usize = 20;

// Static variables
static GLOBAL_SERVICE: Option<UserService> = None;
static mut DEBUG_MODE: bool = false;

// User struct
pub struct User {
    pub id: u32,
    pub name: String,
    pub email: String,
    pub is_active: bool,
}

// UserService struct
pub struct UserService {
    users: HashMap<u32, User>,
}

// Implementation block for User
impl User {
    // Associated function (constructor)
    pub fn new(id: u32, name: String, email: String) -> Self {
        User {
            id,
            name,
            email,
            is_active: true,
        }
    }

    // Method
    pub fn activate(&mut self) {
        self.is_active = true;
    }

    // Method
    pub fn deactivate(&mut self) {
        self.is_active = false;
    }

    // Method that calls another function
    pub fn is_valid(&self) -> bool {
        validate_email(&self.email) && !self.name.is_empty()
    }

    // Method that calls another method
    pub fn format_name(&self) -> String {
        format_user_name(&self.name)
    }
}

// Implementation block for UserService
impl UserService {
    // Associated function (constructor)
    pub fn new() -> Self {
        UserService {
            users: HashMap::new(),
        }
    }

    // Method
    pub fn add_user(&mut self, user: User) -> Result<(), String> {
        if self.users.contains_key(&user.id) {
            return Err(format!("User {} already exists", user.id));
        }
        self.users.insert(user.id, user);
        Ok(())
    }

    // Method
    pub fn get_user(&self, id: u32) -> Option<&User> {
        self.users.get(&id)
    }

    // Method
    pub fn update_user(&mut self, user: User) -> Result<(), String> {
        if !self.users.contains_key(&user.id) {
            return Err(format!("User {} not found", user.id));
        }
        self.users.insert(user.id, user);
        Ok(())
    }

    // Method
    pub fn delete_user(&mut self, id: u32) -> Result<(), String> {
        self.users.remove(&id).map(|_| ()).ok_or_else(|| format!("User {} not found", id))
    }

    // Method that calls another method
    pub fn list_active_users(&self) -> Vec<&User> {
        self.users.values().filter(|user| user.is_active).collect()
    }

    // Method that calls another function
    pub fn get_user_by_email(&self, email: &str) -> Option<&User> {
        self.users.values().find(|user| user.email == email)
    }
}

// Standalone function
pub fn validate_email(email: &str) -> bool {
    email.contains('@') && email.contains('.')
}

// Standalone function
pub fn format_user_name(name: &str) -> String {
    name.trim().to_uppercase()
}

// Standalone function that calls other functions
pub fn create_user(id: u32, name: String, email: String) -> Result<User, String> {
    if !validate_email(&email) {
        return Err(format!("Invalid email: {}", email));
    }

    let formatted_name = format_user_name(&name);

    Ok(User {
        id,
        name: formatted_name,
        email,
        is_active: true,
    })
}

// Standalone function that calls methods
pub fn process_users(service: &mut UserService, users: Vec<User>) -> Result<(), String> {
    for user in users {
        service.add_user(user)?;
    }
    Ok(())
}

// Enum example
pub enum UserRole {
    Admin,
    User,
    Guest,
}

// Enum with associated data
pub enum UserStatus {
    Active,
    Inactive,
    Suspended { reason: String },
}

// Trait definition
pub trait Validator {
    fn validate(&self) -> Result<(), String>;
}

// Trait implementation for User
impl Validator for User {
    fn validate(&self) -> Result<(), String> {
        if self.name.is_empty() {
            return Err("Name cannot be empty".to_string());
        }
        if !validate_email(&self.email) {
            return Err("Invalid email".to_string());
        }
        Ok(())
    }
}

// Generic function
pub fn find_user_by<T, F>(users: &[User], predicate: F) -> Option<&User>
where
    F: Fn(&User) -> Option<T>,
{
    users.iter().find_map(predicate)
}

// Function that uses the generic function
pub fn find_user_by_id(users: &[User], id: u32) -> Option<&User> {
    find_user_by(users, |user| if user.id == id { Some(user) } else { None })
}

// Point struct for testing
pub struct Point {
    pub x: f64,
    pub y: f64,
}

// Implementation for Point
impl Point {
    pub fn new(x: f64, y: f64) -> Self {
        Point { x, y }
    }

    pub fn distance(&self) -> f64 {
        (self.x * self.x + self.y * self.y).sqrt()
    }
}

// Main function
pub fn main() {
    let mut service = UserService::new();

    let user1 = create_user(1, "John Doe".to_string(), "john@example.com".to_string()).unwrap();
    let user2 = create_user(2, "Jane Smith".to_string(), "jane@example.com".to_string()).unwrap();

    service.add_user(user1).unwrap();
    service.add_user(user2).unwrap();

    let active_users = service.list_active_users();
    println!("Active users: {}", active_users.len());
}

